Add database object and database object import rule types

Tener opened the pull request:
Two new are resources added: database objects and database object import rules. Both resources follow guidelines from [RFD 153](https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md). In the first phase of implementation, the database objects will be ephemeral, which means the corresponding RPC service is not needed at this point. This will change in the second phase, which will deal with TAG integration.

This is a PR in a series of PRs implementing [RFD 151](https://github.com/gravitational/teleport/blob/master/rfd/0151-database-permission-management.md).

Tener committed:
Add database object and database object import rule types.

Tener commented:
To see how the change looks like in its entirety, please see: https://github.com/gravitational/teleport/pull/35034.

greedy52 commented on api/types/databaseobject/databaseobject.go:
```diff
@@ -0,0 +1,75 @@
+	if obj.Metadata == nil {
+		return trace.BadParameter("metadata: must be non-nil")
+	}
+	if obj.Metadata.Name == "" {
+		return trace.BadParameter("metadata.name: must be non-empty")
+	}
```
`obj.Metadata.CheckAndSetDefaults()`?

codingllama replied:
I didn't find anything either.

greedy52 commented on api/types/databaseobject/databaseobject.go:
```diff
@@ -0,0 +1,75 @@
+	return databaseObject, nil
+}
+
+// ValidateDatabaseObject checks if *dbobjectv1.DatabaseObject is valid.
+func ValidateDatabaseObject(obj *dbobjectv1.DatabaseObject) error {
```
Any advantage to having this `package.ValidateXXX` function vs the common `obj.CheckAndSetDefaults`? Do we expect the function to be used other places?

Tener replied:
We can't have `CheckAndSetDefaults()` or any other custom methods in the RFD 153 world.

greedy52 replied:
I didn't catch that part. Let me read through it again =p

Tener replied:
I'm not sure if this is explicitly mentioned in the RFD, but that is the reality right now. We could, in theory, write a custom codegen for subset of our message types, but there is a resistance to this notion: codegens are somewhat hard to maintain.

Tener commented on api/proto/teleport/dbobject/v1/dbobject.proto:
```diff
@@ -0,0 +1,52 @@
+  string protocol = 1;
+  // The name of the database service that this object belongs to.
+  string db_service_name = 2;
+  // The kind of database object (e.g., table, view).
+  string object_kind = 3;
```
I didn't consider that. An interesting idea, but let me think about this further.

codingllama replied:
My 2c is that it's better to keep this separate, as we could use sub_kind to further distinguish "object kinds" from one another.

Tener replied:
I like Alan's reasoning, and I agree. Let's keep this unpopulated for now, with an option of using this in the future.

Tener commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto:
```diff
@@ -0,0 +1,82 @@
+  // priority represents the priority of the rule application. Rules with higher priorities override those with lower.
+  int32 priority = 1;
+
+  // db_labels is a set of labels matched against database labels.
+  map<string, string> db_labels = 2;
```
Sounds reasonable and shouldn't be too hard to change, I'll go with it.

Tener commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto:
```diff
@@ -0,0 +1,82 @@
+  // table_names specify the names of the tables to match.
+  repeated string table_names = 1;
+
+  // view_names specify the names of the views to match.
+  repeated string view_names = 2;
+
+  // procedure_names specify the names of the procedures to match.
+  repeated string procedure_names = 3;
```
Well, this is true, although the implementation is a little forward-looking in this aspect, partially to make sure this is a dimension we can expand it easily.

greedy52 replied:
Cool. If we leave them, maybe consider throwing an error when they have non-empty values.

codingllama commented on api/proto/teleport/dbobject/v1/dbobject.proto:
```diff
@@ -0,0 +1,52 @@
+  // The kind of resource represented.
+  string kind = 1;
+  // Differentiates variations of the same kind. All resources should
+  // contain one, even if it is never populated.
```
Do database objects have sub-kinds? That is more interesting to know than the generic blurb.

Tener replied:
Done.

codingllama commented on api/proto/teleport/dbobject/v1/dbobject.proto:
```diff
@@ -0,0 +1,52 @@
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobject/v1;dbobjectv1";
+
+// DatabaseObject represents a database object that can be imported into Teleport.
```
What is database object? An example could go a long way. Please document.

Tener replied:
This is and other pieces of information are covered in depth in the RFD. I'd rather keep a single source of truth here, but I can see an argument for adding a little more detail to the godocs.

Tener replied:
Done.

codingllama replied:
That is already very helpful, thanks.

codingllama commented on api/proto/teleport/dbobject/v1/dbobject.proto:
```diff
@@ -0,0 +1,52 @@
+
+// DatabaseObjectSpec is the spec for the database object.
+message DatabaseObjectSpec {
+  // The protocol used to connect to the database (e.g., postgres).
```
Are all fields required? Please document on which field whether it's required or optional.

Tener replied:
Is this really a relevant information to put here? What makes an object valid tends to be complex, often only easily express as a code. Just take a look on `message RoleConditions` and how it doesn't speak about validation for each field, while in practice the rules are in place, and pretty complex too.

codingllama replied:
> Is this really a relevant information to put here?

Yes, I think so. I'm not asking you to express complex scenarios, although arguably these are even more valuable to document. If it's simple to say whether a field is optional or required, that is helpful information for users and readers of your API. Think of it as part of the contract.

For example, take a look at the last comment line in every field here: https://github.com/gravitational/teleport/blob/30e0e2f1baf83c8ae3f08999be7aa6b27ecbdcc2/api/proto/teleport/devicetrust/v1/device.proto#L41-L43

This is simple to do and clearly documents the API expectations.

codingllama commented on api/proto/teleport/dbobject/v1/dbobject.proto:
```diff
@@ -0,0 +1,52 @@
+  // The protocol used to connect to the database (e.g., postgres).
+  string protocol = 1;
+  // The name of the database service that this object belongs to.
+  string db_service_name = 2;
```
nit: avoid abbreviations in public APIs.

database_service_name or just service_name would both work.

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto:
```diff
@@ -0,0 +1,82 @@
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobjectimportrule/v1;dbobjectimportrulev1";
+
+// DatabaseObjectImportRule is the resource representing a global database object import rule.
```
What is an import rule? What is it used for? Please document.

Tener replied:
Done.

codingllama replied:
Much better, thanks!

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto:
```diff
@@ -0,0 +1,82 @@
+  // The kind of resource represented.
+  string kind = 1;
+  // Differentiates variations of the same kind. All resources should
+  // contain one, even if it is never populated.
```
Same as the other, does it have any valid sub_kinds?

Tener replied:
No, but it is here for a common interface with other resource types.

Tener replied:
I've updated the description.

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto:
```diff
@@ -0,0 +1,82 @@
+
+// DatabaseObjectImportRuleSpec is the spec for database object import rule.
+message DatabaseObjectImportRuleSpec {
+  // priority represents the priority of the rule application. Rules with higher priorities override those with lower.
```
What is the scale for priorities? Is zero the highest or the bigger number the highest? (It seems like the latter but let's be 100% clear.)

Is zero a valid number?

Are negative numbers valid priorities?

Tener replied:
Yes, all integers are valid. How priorities are processed is explained in the context of the whole process, which is way too long to repeat in this place, but covered in RFD.

codingllama replied:
Could we have a brief explanation here too? Pointing people to the RFD is fine, but it doesn't hurt saying whether negative is valid, or whether the highest priority is the highest number. We have more RFDs than most of us can keep up with, so let's try to make simple details avaible.

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto:
```diff
@@ -0,0 +1,82 @@
+
+// DatabaseObjectImportRuleSpec is the spec for database object import rule.
+message DatabaseObjectImportRuleSpec {
+  // priority represents the priority of the rule application. Rules with higher priorities override those with lower.
```
Are all fields required? Please document on which field whether it's required or optional.

Tener replied:
See my other comment.

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto:
```diff
@@ -0,0 +1,82 @@
+  DatabaseObjectImportScope scope = 2;
+
+  // AddLabels specifies which labels to add if the match succeeds.
+  map<string, string> add_labels = 3;
```
Suggestion: labels_to_add.

It's a bit more explicit and starts with the noun. Plays well if we ever want to add labels_to_remove, labels_to_update, etc.

codingllama replied:
Gotcha.

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule.proto:
```diff
@@ -0,0 +1,82 @@
+// DatabaseObjectImportScope specifies the object scope.
+message DatabaseObjectImportScope {
+  // database_names specify the names of the databases to include in the scope.
+  repeated string database_names = 1;
```
If there are multiple database_names and schema_names, what is the final match? The combination of all databases and schemas? Please document.

Is there a way to represent database+schema in a way that it wouldn't create a combinatory explosion? Or is the answer to that "do multiple mappings" ?

I guess what I'm trying to say is: is making these fields repeated a reasonable design choice, or should we change them to non-repeated and require multiple mappings instead?

Tener replied:
Please see the RFD for the details. Given that RFD is commited to the repo, I consider it part of the project documentation I can refer people too; I'll add a note to read it for details behind these data types.

Tener replied:
I've added a note to go to RFD for details.

codingllama replied:
Please document the answer to the first question, it won't take much of your time and it will be helpful to everyone who is not yet in the know and is trying to make sense of this protos, like myself and other reviewers.

I would appreciate if you could answer the other two questions, even if just in this comment thread, or maybe point me to a specific part of the RFD that addresses this.

Tener replied:
> If there are multiple database_names and schema_names, what is the final match? The combination of all databases and schemas? Please document.

You don't care which particular entry in database_names did you match, same for schema_names. Overall, the result of a match is boolean.

> Is there a way to represent database+schema in a way that it wouldn't create a combinatory explosion? Or is the answer to that "do multiple mappings" ?

DatabaseObjectImportScope is isomorphic with `([]string, []string)` pair, rather than `[](string, string)`.We simply match against both fields independently, with wildcard `*` support.

Here is the actual code (planned; can't have it on master without merging its dependencies first):

https://github.com/gravitational/teleport/blob/c0afc5d8b642a4e68c4a5f99c96c52383f382ccf/lib/srv/db/common/permissions/import_rules.go#L99-L116

> I guess what I'm trying to say is: is making these fields repeated a reasonable design choice, or should we change them to non-repeated and require multiple mappings instead?

I believe this is a sound design that avoids needless repetition, but I'm open to counter examples in which a different design could be better.

codingllama replied:
OK, so it does match the combination of all databases+schemas.

> I believe this is a sound design that avoids needless repetition, but I'm open to counter examples in which a different design could be better.

👍

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule_service.proto:
```diff
@@ -0,0 +1,97 @@
+// The request for CreateDatabaseObjectImportRule.
+message CreateDatabaseObjectImportRuleRequest {
+  // The DatabaseObjectImportRule to create.
+  DatabaseObjectImportRule item = 1;
```
Maybe:

```suggestion
  DatabaseObjectImportRule rule = 1;
```

Tener replied:
I actually prefer `item`, but `rule` works too.

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule_service.proto:
```diff
@@ -0,0 +1,97 @@
+// The response for ListDatabaseObjectImportRules.
+message ListDatabaseObjectImportRulesResponse {
+  // The page of DatabaseObjectImportRules that matched the request.
+  repeated DatabaseObjectImportRule items = 1;
```
Suggestion:

```suggestion
  repeated DatabaseObjectImportRule rules = 1;
```

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule_service.proto:
```diff
@@ -0,0 +1,97 @@
+// The request for UpdateDatabaseObjectImportRule.
+message UpdateDatabaseObjectImportRuleRequest {
+  // The DatabaseObjectImportRule to replace.
+  DatabaseObjectImportRule item = 1;
```
```suggestion
  DatabaseObjectImportRule rule = 1;
```

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule_service.proto:
```diff
@@ -0,0 +1,97 @@
+// The request for UpsertDatabaseObjectImportRule.
+message UpsertDatabaseObjectImportRuleRequest {
+  // The DatabaseObjectImportRule to create or replace.
+  DatabaseObjectImportRule item = 1;
```
```suggestion
  DatabaseObjectImportRule rule = 1;
```

codingllama commented on api/proto/teleport/dbobjectimportrule/v1/dbobjectimportrule_service.proto:
```diff
@@ -0,0 +1,97 @@
+}
+
+// The request for UpdateDatabaseObjectImportRule.
+message UpdateDatabaseObjectImportRuleRequest {
```
No update_mask on Update? I recommend you add one, it helps to prevent squashing fields we didn't want to touch in the first place.

* https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md#update
* https://cloud.google.com/apis/design/standard_methods#update

Tener replied:
As far as I can see `update_mask` leads to complex, brittle code that requires a ton of testing for little benefit. For example, this string comparisons is nasty:
```golang
		case path == "spec.roles":
			role.SetImpersonateConditions(types.Allow, types.ImpersonateConditions{
				Roles: req.Bot.Spec.Roles,
			})
```

It is dangerous to use, too: looking at other types using it, the clients would typically calculate the "patch" that needs applying and send it over; only the "patch" is calculated based off their own view of the object state, which may be stale. To be safe, you should do db-style transactions with retries, where client logic only operates on a consistent view of the object.

This kind of call makes more sense for editing UIs where the person performing the edits doesn't care that unrelated parts of the object has been changed, although even then it should be handled with care.

Tener replied:
@strideynet what are your thoughts on `update_mask` after implementing necessary changes for bots?

strideynet replied:
They can be a tad repetitive to implement - but, the benefits have been fairly clear. i.e the frontend no longer needs to make a Get before it Updates the Bot when looking to simply update the roles.

I think realistically long term we'd be wanting to investigate something that can leverage proto message reflection to automatically apply the update mask.

Tener replied:
Cool, makes sense in the context of UI, thanks for the context. Does it properly handle the case where two people are editing the set of roles for the bot?

strideynet replied:
Not quite - but I wouldn't take the Bot resource as an example here. It's not persisted to the backend at the moment and is on-the-spot reconciled to and from a Bot User and Bot Role - this isn't the long-term plan but is part of a transition as historically we haven't had a Bot resource (until v15). This somewhat limits my ability to use the current optimistic locking.

codingllama replied:
Update masks are good in situations where not every update is a full update. I find that this happens frequently.

I'm a bit torn about calculating the mask on the fly, as that somewhat defeats the advantage. Fixed masks work better in my opinion (as in, I know this will always and only ever touch fields _x_ and _y_, so these are the masks used).

Another advantage is that you may not need a previous read, like Noah said.

> Does it properly handle the case where two people are editing the set of roles for the bot?

That is not what the masks are for, although they can help "solve" those problems if the concurrent edits are independent and touch disjointed fields. I'd look at optimistic concurrency in this case, which is something that we also miss in Teleport designs.

codingllama replied:
Just to be clear about expectations, I'm not holding the PR for the update_mask, but I would strongly recommend you added it (for whatever that is worth).

Tener replied:
I gave it some further thought. While I recognise the potential benefits of `update_mask`, these are irrelevant to the current needs of this API. It is a kind of optimisation, which I believe to be premature at this moment. I'm certainly open to revisiting this in the future if needed; the request can be easily extended to accommodate that.

As for the general idea of `update_mask`, it would be quite interesting to see an end-to-end implementation which accounts for the complexity I've mentioned above.

codingllama commented on api/types/databaseobject/databaseobject.go:
```diff
@@ -0,0 +1,75 @@
+		Spec: spec,
+	}
+
+	err := ValidateDatabaseObject(databaseObject)
```
I'm of the strong opinion that validation is a storage responsibility. This should be gated by storage during Create/Update/Upsert and not whenever a new resource is initialized.

This pattern bakes validation into clients and into multiple layers, diffusing responsibility and making validation changes difficult to do.

* https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md#resource-validation

Tener replied:
The linked sections says:

> Resource validation
The strictest validation of a resource should be performed prior to write operations. Any resource persisted in the backend should be guaranteed to be valid. Read operations should not perform resource validations, doing so could prevent a resource being read if validations are modified to be more restrictive after a resource had already been written. This allows for resilient reads which ensure that any resource stored is always allowed to be retrieved from the backend.

It does not say that no validation should occur outside of storage. It does the opposite, really: using the word "strictest" suggest some others are less strict.

More generally, performing validation on the write path doesn't help with usage of ephemeral objects. It is a common pattern to validate an object on creation, i.e. in `NewFoo()` function for type `Foo`. This only makes sense: you want rapid feedback if you are trying to create an invalid object.

Now, I can see an argument for creating two tiers of validation, one more strict than the other, and only using the strict form on writes. But that sounds like unnecessary complexity, as you now have to carefully maintain two distinct sets of requirements for no clear benefit.

codingllama replied:
The decider of what "valid" means is typically the backend. Pushing validation to New functions can make clients become out of sync with the backends and over-validate. It can also cause problems when reading stored values that "New" considers invalid (but should be treated resiliently instead).

Agree to disagree.

codingllama commented on api/types/databaseobject/databaseobject.go:
```diff
@@ -0,0 +1,75 @@
+// ValidateDatabaseObject checks if *dbobjectv1.DatabaseObject is valid.
+func ValidateDatabaseObject(obj *dbobjectv1.DatabaseObject) error {
+	if obj == nil {
+		return trace.BadParameter("object must be non-nil")
```
```suggestion
		return trace.BadParameter("database object must be non-nil")
```

Just "object" could be about anything, which can make this error confusing if it's coming from several layers down.

Similarly for other errors.

Tener replied:
Isn't this the reason for using `trace` package in the first place? You are getting a full stack trace information, which gives you this information and more.

Tener replied:
I've changed the message to indicate the object type throwing the error.

codingllama commented on api/types/databaseobject/databaseobject.go:
```diff
@@ -0,0 +1,75 @@
+		return trace.BadParameter("metadata.name: must be non-empty")
+	}
+	if obj.Kind != types.KindDatabaseObject {
+		return trace.BadParameter("wrong kind %v", obj.Kind)
```
```suggestion
		return trace.BadParameter("invalid kind for database object: %v", obj.Kind)
```

codingllama commented on api/types/databaseobject/databaseobject.go:
```diff
@@ -0,0 +1,75 @@
+
+// ValidateDatabaseObject checks if *dbobjectv1.DatabaseObject is valid.
+func ValidateDatabaseObject(obj *dbobjectv1.DatabaseObject) error {
+	if obj == nil {
```
Suggestion: consider a switch chain, it's idiomatic Go and slightly more compact.

```go
	switch {
	case obj == nil:
		return trace.BadParameter("object must be non-nil")
	case obj.Metadata == nil:
		return trace.BadParameter("metadata: must be non-nil")
	// ... so on
	}
```

Tener replied:
It's a nice hack, but it hurts readability in my view. It is much less popular in our codebase, too, so using this pattern hurts homogeneity of the codebase for no strong reason.

codingllama commented on api/types/databaseobject/databaseobject_test.go:
```diff
@@ -0,0 +1,148 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			obj, err := NewDatabaseObject(tt.name, tt.spec)
+			require.ErrorIs(t, err, tt.expectedError)
```
Please avoid comparing complex errors with ErrorIs. These comparisons are brittle and break easily.

codingllama replied:
ErrorAs is much more resilient if you want to check types. trace.IsBadParameter and like methods are similarly resilient.

ErrorContains is less brittle if you want to check that you are hitting a particular error message (say, in cases where you have many BadParameter variants).

codingllama replied:
(Cont, sent too soon)

ErrorsIs is good for guard errors like io.EOF and the like - ie, cases where the entire instance of an error is guaranteed to appear as-is.

Tener replied:
Agreed about preferred use of different helpers; in this case, I want to check that I'm hitting very particular error condition with my test case, which is why I care both about the error type and message. That makes me believe `require.ErrorIs` is the right choice.

codingllama commented on api/types/databaseobject/databaseobject_test.go:
```diff
@@ -0,0 +1,148 @@
+	"github.com/gravitational/teleport/api/types"
+)
+
+func TestNewDatabaseObject(t *testing.T) {
```
Up to you, but I would be fine with this test not existing.

Tener replied:
If I can get additional coverage for my change with little effort, in general I try to do so. In the context of this PR there are no other callers of `NewDatabaseObjectImportRule` other than this test, so dropping this test would leave it entirely without coverage (for now).

codingllama commented on api/types/databaseobject/databaseobject_test.go:
```diff
@@ -0,0 +1,148 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := ValidateDatabaseObject(tt.databaseObject)
+			require.ErrorIs(t, err, tt.expectedError)
```
Ditto for ErrorsIs. I suggest a combination of ErrorContains and either ErrorAs(new(*trace.BadParameterError)) or a trace.IsBadParameter check.

Tener replied:
I disagree this is a problem, see my argument in the other comment.

codingllama commented on api/types/databaseobjectimportrule/databaseobjectimportrule.go:
```diff
@@ -0,0 +1,75 @@
+		Spec: spec,
+	}
+
+	err := ValidateDatabaseObjectImportRule(out)
```
Same remark about validation to storage.

Tener replied:
Same counter argument 😉

codingllama commented on api/types/databaseobjectimportrule/databaseobjectimportrule.go:
```diff
@@ -0,0 +1,75 @@
+// ValidateDatabaseObjectImportRule checks if dbobjectimportrulev1.DatabaseObjectImportRule is valid.
+func ValidateDatabaseObjectImportRule(rule *dbobjectimportrulev1.DatabaseObjectImportRule) error {
+	if rule == nil {
+		return trace.BadParameter("must be non-nil")
```
Ditto for ambiguous error.

```suggestion
		return trace.BadParameter("database object import rule be non-nil")
```

Tener replied:
You still have the stack trace, but it doesn't cost much to improve this, so done.

Tener replied:
I've changed the message.

codingllama commented on api/types/databaseobjectimportrule/databaseobjectimportrule_test.go:
```diff
@@ -0,0 +1,172 @@
+	"github.com/gravitational/teleport/api/types"
+)
+
+func TestNewDatabaseObjectImportRule(t *testing.T) {
```
Ditto for this test not needing to exist.

Tener replied:
Disagreed, as per the other comment.

codingllama commented on api/types/databaseobjectimportrule/databaseobjectimportrule_test.go:
```diff
@@ -0,0 +1,172 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewDatabaseObjectImportRule(tt.name, tt.spec)
+			require.ErrorIs(t, err, tt.expectedError)
```
Please avoid ErrorIs.

Tener replied:
I think it's fine here, see other comment.

codingllama commented on api/types/databaseobjectimportrule/databaseobjectimportrule_test.go:
```diff
@@ -0,0 +1,172 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := ValidateDatabaseObjectImportRule(tt.rule)
+			require.ErrorIs(t, err, tt.expectedError)
```
Please avoid ErrorIs. Same recommendation for ErrorContains and ErrorAs or IsBadParameter.

Tener committed:
Address review feedback.

- Switch to multiple allowed values per label.
- Improve some error messages
- Avoid abbreviated where possible.

Tener committed:
Add context for resource types

Tener committed:
Update comment for sub_kind fields.

Tener committed:
Update test to accomodate new field name.

Tener committed:
Fix tests

Tener committed:
item => rule

Tener committed:
Extend proto docs

Tener commented:
Thanks for the reviews, folks!
